File: solutions/typescript/phone-number/1/phone-number.ts
export function clean(phoneNumber: string): string {
    if (phoneNumber.match(/[a-zA-Z]/)) {
        throw new Error('Letters not permitted');
    }

    if (phoneNumber.match(/[@:!]/)) {
        throw new Error('Punctuations not permitted');
    }

    const cleanedNumber = phoneNumber.replace(/\s|\(|\)|-|\.|\+/g, '');

    if (cleanedNumber.length < 10) {
        throw new Error('Incorrect number of digits');
    }

    if (cleanedNumber.length > 11) {
        throw new Error('More than 11 digits');
    }

    if (cleanedNumber.length === 11) {
        if (cleanedNumber.startsWith('1')) {
            const numberWithoutCountryCode = cleanedNumber.substring(1);
            validateNumber(numberWithoutCountryCode);
            return numberWithoutCountryCode;
        }
        throw new Error('11 digits must start with 1');
    }

    if (cleanedNumber.length === 10) {
        validateNumber(cleanedNumber);
        return cleanedNumber;
    }

    throw new Error('Invalid number');
}

function validateNumber(number: string): void {
    if (number.charAt(0) === '0') {
        throw new Error('Area code cannot start with zero');
    }

    if (number.charAt(0) === '1') {
        throw new Error('Area code cannot start with one');
    }

    if (number.charAt(3) === '0') {
        throw new Error('Exchange code cannot start with zero');
    }

    if (number.charAt(3) === '1') {
        throw new Error('Exchange code cannot start with one');
    }
}

File: solutions/go/phone-number/1/phone_number.go
package phonenumber

import (
	"errors"
	"fmt"
)

func Number(phoneNumber string) (string, error) {
	return clean(phoneNumber)
}

func AreaCode(phoneNumber string) (string, error) {
	clean, err := clean(phoneNumber)
	if err != nil {
		return "", err
	}
	return clean[:3], nil
}

func Format(phoneNumber string) (string, error) {
	clean, err := clean(phoneNumber)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("(%s) %s-%s", clean[:3], clean[3:6], clean[6:]), nil
}

func clean(phoneNumber string) (string, error) {
	number := []rune{}
	for _, c := range phoneNumber {
		if c >= '0' && c <= '9' {
			number = append(number, c)
		}
	}
	length := len(number)
	if length == 11 {
		if number[0] != '1' {
			return "", errors.New("invalid number")
		}
		number = number[1:]
	}
	if length < 10 || length > 11 {
		return "", errors.New("invalid number")
	}
	if number[0] == '0' || number[0] == '1' {
		return "", errors.New("invalid number")
	}
	if number[3] == '0' || number[3] == '1' {
		return "", errors.New("invalid number")
	}
	return string(number), nil
}
